content/blog/webr-0-3-1/index.md

```diff
@@ -0,0 +1,368 @@
+
+### Graphics device bug fixes
+
+In addition to adding the ability to capture graphics output, the [`webr::canvas()`](https://rdrr.io/pkg/webr/man/canvas.html) graphics device has also had various bug fixes made to better implement R base graphics. The easiest way to demonstrate is probably by example:
+
+<div class="highlight">
+
+{{< webr-editor code=`# The lty and lwd graphical properties now work correctly\nplot(1:10, type = "l", lty = 2, lwd = 3)\npoints(1:10, cex = 3, lwd = 2)` width=504 height=311.472 >}}
+
+</div>
+
+<div class="highlight">
+
+{{< webr-editor code=`# The cex graphical property is now taken into account\n# when calculating font sizes\nplot(1, main = "This is a large title", cex.main = 3)` width=504 height=311.472 >}}
+
+</div>
+
+<div class="highlight">
+
+{{< webr-editor code=`# Rasters with negative width or height are now correctly\n# drawn mirrored and flipped.\ninstall.packages("jpeg")\nlogo = jpeg::readJPEG(system.file(package = "jpeg", "img", "Rlogo.jpg"))\nplot(NULL, xlab = "", ylab = "", xlim = c(0, 1), ylim = c(0, 1))\n\nrasterImage(logo, xleft = 0.2, xright = 0.5, ybottom = 0.5, ytop = 1)\nrasterImage(logo, xleft = 0.8, xright = 0.5, ybottom = 0.5, ytop = 1)\nrasterImage(logo, xleft = 0.2, xright = 0.5, ybottom = 0.5, ytop = 0)\nrasterImage(logo, xleft = 0.8, xright = 0.5, ybottom = 0.5, ytop = 0)` width=504 height=311.472 >}}
+
+</div>
+
+## The R object interface
+
+The R object interface provided by webR has been expanded to support the conversion of more types of JavaScript objects into R objects. Such conversions are automatically applied when interacting with the R environment from JavaScript.
+
+### Raw vectors
+
+JavaScript objects of type `TypedArray`, `ArrayBuffer`, and `ArrayBufferView` (e.g. [`Uint8Array`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array)) may now be used to construct R objects. By default, objects of this type are converted to R raw atomic vectors. This simplifies the transfer of binary data to R.
+
+``` javascript
+const data = new Uint8Array([4, 12, 8, 24, 15, 12]);
+// Print data's R object class and an example byte
+await webR.evalR(`
+  class(data)
+  data[2]
+`, { withAutoprint: true, env: { data } });
+```
+
+<div class="output">
+
+    [1] "raw"
+    [1] 0c
+
+</div>
+
+### R `data.frame`
+
+JavaScript objects of shape `{ x: [...], y: [...] }`, with data in a "long" column-based form, can now be used to construct R objects. In previous versions of webR, this object shape was reserved for future use. However, with this release webR now constructs an R `data.frame` by taking the source object's properties as column vectors. The resulting `data.frame` can then be manipulated from R in the usual way:
```

Comment:
The more I think about it the more I feel this would be better done with a helper ctor that would support both shapes of data frames and would check for data frame invariants on the JS side. Especially since this ctor downgrades to lists silently. It seems to me that the ease of use of this syntax is not worth the very real downsides on programming errors. We are compounding silent coercions and laziness of checking (if any) here.

If this is just for debugging then this should probably not be documented prominently. But also I feel like a short-name ctor would be just as good for interactivity.

Reply:
As discussed in another forum, we'll keep this as-is for now, but address the issues in the next minor release of webR with the aim of stabilising the mapping between JS object shape and R object type.

My current plan is the following changes:

1) Reserve this JS syntax for constructing only `data.frame`s when using the generic `RObject()` constructor.
2) Remove the list fallback and throw an Error instead. R lists can still be constructed using e.g. `RList()`.
3) Add a small `RList` subclass, `RDataFrame`, for explicit `data.frame` construction.
3) Document the above, recommending for production code to use the explicit constructors rather than `RObject`.

We'll continue this discussion in a future PR on the webR repo.